component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -361,24 +366,52 @@ private boolean isValidSignature(String clientId, SignedJWT signedJWT, String te
                                      String alias) throws OAuthClientAuthnException {
 
         X509Certificate cert = null;
+        String jwksUri = "";
+        boolean isValidSignature = false;
         try {
             cert = (X509Certificate) OAuth2Util.getX509CertOfOAuthApp(clientId, tenantDomain);
         } catch (IdentityOAuth2Exception e) {
             if (log.isDebugEnabled()) {
                 log.debug(e.getMessage(), e);
             }
         }
+        //if cert is null check whether a jwks endpoint is configured for the service provider
+        if (cert == null) {
+            try {
+                ServiceProviderProperty[] spProperties = OAuth2Util.getServiceProvider(clientId).getSpProperties();
+                for (ServiceProviderProperty spProperty : spProperties) {
+                    if (Constants.JWKS_URI.equals(spProperty.getName())) {
+                        jwksUri = spProperty.getValue();
+                    }
+                }
+                //validate the signature of the assertion using the jwks end point
+                String jwtString = signedJWT.getParsedString();
+                String alg = signedJWT.getHeader().getAlgorithm().getName();
+                Map<String, Object> options = new HashMap<String, Object>();
+                isValidSignature = new JWKSBasedJWTValidator().validateSignature(jwtString, jwksUri, alg, options);
+
+            } catch (IdentityOAuth2Exception e) {
+                throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
```

Comment:
I think we are missing quite a lot of information here about the cause of the exception. Don't we have a constructor which accepts a throwable in OAuthClientAuthnException ?. If we have, let's use it. We might need to revisit the other places of OAuthClientAuthnException  usage as well. Otherwise we are missing quite a lot of information.

Reply:
used throwable and added error message